File: src/main/python/AoC2020_17.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 17
#

import sys
from typing import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.game_of_life import ClassicRules
from aoc.game_of_life import GameOfLife
from aoc.game_of_life import InfiniteGrid

TEST = """\
.#.
..#
###
"""

ON = "#"
GENERATIONS = 6
CellFactory = Callable[[int, int], tuple[int, ...]]
Input = InputData
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def solve(self, inputs: Input, cell_factory: CellFactory) -> int:
        alive = {
            cell_factory(r, c)
            for r, row in enumerate(inputs)
            for c, state in enumerate(row)
            if state == ON
        }
        game_of_life = GameOfLife(alive, InfiniteGrid(), ClassicRules())
        for i in range(GENERATIONS):
            game_of_life.next_generation()
        return sum(1 for _ in game_of_life.alive)

    def part_1(self, inputs: Input) -> int:
        return self.solve(inputs, cell_factory=lambda r, c: (0, r, c))

    def part_2(self, inputs: Input) -> int:
        return self.solve(inputs, cell_factory=lambda r, c: (0, 0, r, c))

    @aoc_samples(
        (
            ("part_1", TEST, 112),
            ("part_2", TEST, 848),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2020, 17)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/julia/AoC2020_17.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2020_17

include("aoc.jl")
include("game_of_life.jl")

ON = '#'
GENERATIONS = 6

function _parse(input, dim::Int)
    alive = Set([(r, c, 0, 0)
                 for (r, row) ∈ enumerate(input)
                 for (c, state) ∈ enumerate(row)
                 if state == ON])
    type = GameOfLife.InfiniteGrid(dim)
    return GameOfLife.Generation(alive, type, GameOfLife.classicRules)
end

function solve(input, dim::Int)
    gen = _parse(input, dim)
    for i ∈ 1:GENERATIONS
        gen = GameOfLife.next(gen)
    end
    return length(gen.alive)
end

function part1(input)
    return solve(input, 3)
end

function part2(input)
    return solve(input, 4)
end

TEST = @aoc_splitlines("""\
.#.
..#
###
""")

function samples()
    @assert part1(TEST) == 112
    @assert part2(TEST) == 848
end

end # module AoC2020_17

aoc_main(@__FILE__, ARGS, 2020, 17)
